src/lib/project_manager/cocotb.ts

```diff
@@ -0,0 +1,299 @@
+              name: test_name,
+              pass: pass
+            };
+            results.push(test_info);
+          }
+          resolve(results);
+        });
+      } catch (e) {
+        let results = this.get_all_test_fail(tests);
+        resolve(results);
+      }
+    });
+  };
+
+  get_all_test_fail(tests) {
+    let results: {}[] = [];
+    for (let i = 0; i < tests.length; i++) {
+      const test_name = tests[i];
+      let test_info = {
+        name: test_name,
+        pass: false
+      };
+      results.push(test_info);
+    }
+    return results;
+  }
+
+  async get_test_list(makefile_path: string) : Promise<TestItem[]> {
+    try {
+      shell.exec("make",(code, output) => {
+        if (code === 127){
+        vscode.window.showErrorMessage("Install Make to use cocotb");
+        }
+      });
+    } catch (error) {
+      vscode.window.showErrorMessage("Error testing deps."); 
+    }
+
+    let wksp_folder;
+    if(vscode.workspace.workspaceFolders !== undefined) {
+      wksp_folder = vscode.workspace.workspaceFolders[0].uri.fsPath;
+    } else {
+      return([]);
+    }
+
+    // let found_makefiles = glob.sync('/**/Makefile', {root: wksp_folder});
+    let found_makefiles = [makefile_path];
+    let cocotb_makefiles: {makefile: string, modules: [string]}[] = [];
+
+    const cocotb_makefile_must_contain = "include $(shell cocotb-config --makefiles)/Makefile.sim";
+
+    for (let item of found_makefiles) {
+      let data = fs.readFileSync(path_lib.resolve(item));
+      if (data.includes(cocotb_makefile_must_contain)) {
+        var new_data = data.toString().replace(cocotb_makefile_must_contain, '');
+        const tmpobj = tmp.fileSync();
+
+        fs.writeFileSync(tmpobj.name, `${new_data}\nprint-%  : ; @echo $* = $($*)\n`);
+
+        let modules = child_process.execSync(`make -f ${tmpobj.name} print-MODULE`).toString();
+        modules = modules.replace(/(\r\n|\n|\r)/, '');
+        let modules_arr = modules.split(' ');
+        modules_arr = modules_arr.filter(item => item !== "MODULE");
+        modules_arr = modules_arr.filter(item => item !== "=");
+        cocotb_makefiles.push({makefile: item, modules: modules_arr});
+      }
+    }
+
+    let test_array: TestItem[] = [];
+
+    for (let makefile of cocotb_makefiles) {
+      for (let module of makefile.modules) {
+        const test_name = module;
+        const module_path = path_lib.dirname(makefile.makefile);
+        const filename = path_lib.join(module_path, test_name);
+
+        let python_cocotb_data;
+        try {
+          python_cocotb_data = fs.readFileSync(`${filename}.py`);
+        } catch {
+          vscode.window.showErrorMessage(`Found "${module}" module in Makefile's MODULE variable but no "${test_name}.py" file found`);
+          continue;
+        }
+        
+        let input = python_cocotb_data.toString();
+        let regex = /(^[^#\r\n]* *@cocotb\.test\(\)(\r\n|\n|\r))([^#\r\n]* *(async )?def ([\w_]+)\([\w_]+\):$)/gm;
```

Comment:
It was due to the 'regex' that I couldn't detect some of my test. I tried with the following regex and worked for all. But I don't have much experience with cocotb. 
```js
let regex = /(^[^#\r\n]* *@cocotb\.test\(([^)]*)\)(\r\n|\n|\r))([^#\r\n]* *(async )?def +([\w_]+) *\([\w_]+\):$)/gm;
```
What do you think @ila-embsys ?

Reply:
Sure! I did not take into account that decorator can have arguments.

But the part below is not cover all possible variants of Python lexical yet.
```Regex
\(([^)]*)\)
```
For example, how it can look:

```Python
@cocotb.test(skip = cocotb.SIM_NAME.lower().startswith("icarus"))
async def test_case(dut):
```

So, we should cover parentheses and others.

I propose something like:

```Regex
\([ \t\w=.()"'\[\],]*\)
```

It looks a little creepy but it seems still valid.

Also, I noticed that in regex added detection of extra whitespace. That right. Python allows whitespace and tabs between tokens. For example:
```Python
async 	def 	test_case 	( 	dut 	) 	: 
```

So I propose to add more ```[ \t]+``` between tokens.

Finally, I got the next regexp: 😬

```Regexp
(^[^#\r\n]* *@cocotb\.test[ \t]*\([ \t\w=.()"'\[\],]*\)[ \t]*(\r\n|\n|\r))([^#\r\n]*[ \t]*(async)?def[ \t]+([\w_]+)[ \t]*\([ \t\w_]+\)[ \t]*:[ \t]*$)
```

Is it still acceptable @smgl9 ?

Reply:
looks fine, let's go ahead with it :)